Rewrite core in Typescript

tcbyrd opened the pull request:
This PR is an initial spike at moving the library files to Typescript and is very much a Work in Progress (this is my first time writing Typescript in any real way). Currently all this does is convert the default plugin that renders the Express app to `default.ts` and adds the Express request/response types. Something that is really nice is that even though I had to `import * as express` in the typescript file, since it was only used for adding types, it's not used in the resulting compiled output. Pretty nifty.

In addition to testing against the default plugin, I've used the compiled output to run a couple of other plugins and it seems to be working fine so far. After working through the process and looking at a few other TS projects, I feel like the next steps look like this:

 - [x] Convert the rest of the built-in plugins to `.ts` files and add types
 - [x] Add [@types/bunyan](https://www.npmjs.com/package/@types/bunyan) and convert the logger, middleware, and serializers
 - [x] Since [bottleneck](https://github.com/SGrondin/bottleneck) has type definitions, use that to help convert the GitHub client
 - [x] Convert the helper files like `private-key.js` & `resolver.js`

After that, I feel like I'd have a better understanding of TypeScript and what to do with it before tackling things like `robot.js` and `context.js`.

Of course anyone is welcome to provide feedback or ideas on this process! Especially anyone with more experience with Typescript libraries.

I definitely think this is going to help with serverless/lambda support. The compiled output feels much more like "normal" Javascript that's just been written in an older syntax. I kinda wish I would've started this process earlier! 😄

ocombe commented:
You can use `ts-jest` to test ts files with jest, change the config in package.json to something like that:

```
"jest": {
    "setupFiles": [
      "<rootDir>/test/setup.js"
    ],
    "moduleFileExtensions": [
      "ts",
      "js"
    ],
    "transform": {
      ".+\\.tsx?$": "<rootDir>/node_modules/ts-jest/preprocessor.js"
    },
    "testMatch": [
      "**/test/*.(ts|js)"
    ]
  }
```

tcbyrd commented:
@ocombe thanks for the help! I was actually wondering about this when I was running the tests locally. When working in TS, do you typically only run the tests against the TypeScript files and trust the compiler output, or do you also run tests against the JavaScript files in the resulting build/dist folder?

JasonEtco commented:
cc @wilhelmklopp who's been fiddling with TS testing in a Probot App

ocombe commented:
With ts-jest it precompiles the files, so your tests are running on js files anyway. No need to duplicate the tests.

@wilhelmklopp you can take a look at my app that is in ts: https://github.com/angular/github-robot
I just added some tests.

ocombe commented:
PR to update your branch here: https://github.com/probot/probot/pull/375

tcbyrd commented:
FYI, so I can do some learning in parallel on npm and better understand how this will work when published as a transpiled package, I've published this branch to npm as [probot-ts](https://www.npmjs.com/package/probot-ts).

Seems to work great so far. I've deployed using AWS CLI, wrote a Lambda bot from scratch inside the c9.io IDE, and deployed using the Serverless framework without added plugins using that published package.

tcbyrd commented:
@ocombe I converted the other 2 default plugins. The one for Sentry was pretty straight forward, but I'd definitely like your input on [stats.ts](https://github.com/probot/probot/pull/372/files#diff-1bfed1643ac5a046af57a24905e17162). That one hurt my brain a bit since it has a big  async function that returns an array of objects. I got it to the point of no more compiler warnings and all tests pass, but would appreciate some more experienced 👀 eyes on it. I also published it to the probot-ts npm package as 4.0.1-typescript

ocombe commented on src/plugins/stats.ts:
```diff
@@ -1,9 +1,9 @@
 
   // Cache of stats that get reported
-  const stats = {installations: 0, popular: []}
+  let stats = {installations: 0, popular: [{}]}
```
why are you changing this to let?

tcbyrd replied:
I think I was fighting with the `never[]` type and did this out of confusion. Will revert this.

ocombe commented on src/plugins/stats.ts:
```diff
@@ -1,9 +1,9 @@
 // Built-in plugin to expose stats about the deployment
-module.exports = async robot => {
+module.exports = async (robot: any): Promise<void> => {
```
we should use a type for robot, instead of any

tcbyrd replied:
Since the `robot` type is probably going to be used everywhere, I'm imagining we'll import it into the plugins once we define it in core. Does that sound like the right way to tackle that?

ocombe commented on src/plugins/stats.ts:
```diff
@@ -31,8 +31,8 @@ module.exports = async robot => {
 
-  async function popularInstallations (installations) {
-    let popular = await Promise.all(installations.map(async installation => {
+  async function popularInstallations (installations): Promise<Array<PopularType>> {
```
`PopularType[]` instead of `Array<PopularType>`
add a type for `installations`, and define that type as the type returned by the function `getInstallations`, if typescript gives you trouble in getInstallations, you can return the last line like this: `return github.paginate(req, res => res.data) as InstallationsType;` with `InstallationsType` the new type that you'll define (don't use that name, it's bad)

The `refresh` method should have types as well

ocombe commented on src/plugins/stats.ts:
```diff
@@ -31,8 +31,8 @@ module.exports = async robot => {
-  async function popularInstallations (installations) {
-    let popular = await Promise.all(installations.map(async installation => {
+  async function popularInstallations (installations): Promise<Array<PopularType>> {
+    let popular: PopularType[] = await Promise.all<PopularType>(installations.map(async (installation) => {
```
you don't need to add types here, if you defined a type for `installations` above, it will be inferred by TS, but it's still good for readability to set the type of the variables that you define, so you can leave `let popular: PopularType[]`, but you should remove `<PopularType>` since manually casting a function will hide errors that TS would find otherwise

ocombe commented:
Ok I added a review, I can make a PR if you want

tcbyrd commented:
Awesome thanks for that! I think I'm starting to understand the semantics a little better. I pushed some of those changes.

What types would we add for `refresh()`? It's there to refresh the `stats` variable in memory, so it doesn't explicitly return anything. It's implicitly showing `Promise<void>`, which I thought would be appropriate in that scenario.

tcbyrd commented:
@ocombe I added a few more commits to get this branch up to date with 5.0 and convert the logging middleware. If you have the time I'd ❤️ to get some more experienced eyes 👀 on what I've done so far. ☺️

philschatz commented:
👋  Howdy!

I wanted to build on this branch to help finish converting the files but got confused with some of the files in the diff. I took your changes to `package.json` and ended up converting almost all of the files and documented some snags I ran into along the way. 

I was not sure if I should submit a Pull Request so instead here is a link to the diff: https://github.com/probot/probot/compare/master...philschatz:typescript .

## Snags Encountered

### ... in `lib/github.js`

This code used several non-public methods so I added interfaces to describe them.

### ... in `lib/plugins/stats.js`

[@octokit/rest.js](https://github.com/octokit/rest.js) does not support strongly typed responses; they are all `Promise<Github.AnyResponse>`. 

This was an issue for the `desktop/desktop` group as well (see [desktop/desktop#1937](https://github.com/desktop/desktop/pull/1937) and various discussions at https://github.com/desktop/desktop/issues?q=philschatz)

I would be happy to add those response types to [@octokit/rest.js](https://github.com/octokit/rest.js) since my soon-to-be-abandoned library [octokat.js](https://github.com/philschatz/octokat.js) extends the same [routes.json](https://github.com/philschatz/octokat.js/pull/176/files#diff-c5fbeee2dd3ee01ae5917539dde6afc0)  as [the one @octokit/rest.js uses](https://github.com/octokit/rest.js/blob/master/lib/routes.json)

This was the main reason I was unable to convert the `lib/plugins/stats.js` file to TypeScript.

### The `Logger` type

I was not sure whether to use the Logger type from `bunyan` or use the extended `./lib/logger.js`; I arbitrarily used the bunyan but it is simple to switch.

### Converting tests to Typescript

I did not switch the tests because I was unsure if there was interest in converting them.

tcbyrd commented:
@philschatz Thanks so much for your contribution! I plan on dedicating some time to looking through this today. No comments yet, other than I'm so grateful that people experienced in TS are helping this PR evolve. ✨ ❤️

tcbyrd commented:
@ocombe @philschatz I've made some progress on this PR. Big thanks to both of you for your contributions!! ✨ ✨

I fixed all the merge conflicts and updated some tests to ensure everything works locally, but I'm hitting a snag with the Travis build and was wondering if either of you had some input and getting it working. The build times out, even though it looks like it's passing all the tests. These same test pass locally, and I ran the same commands on a local node docker image and it passes everything there, too. Any thoughts on why Travis seems to not handle this build would be welcome.

zeke commented on .travis.yml:
```diff
@@ -3,22 +3,23 @@ sudo: false
+
 before_install:
-  - npm install -g npm@5
+  - npm install -g typescript
```
Hi! Drive-by commenter. Shouldn't these just be in devDependencies?

tcbyrd replied:
Ah, I think you're right! Travis should just use the local typescript in node_modules. I forgot it was already defined in devDependencies.

ocombe commented on src/context.ts:
```diff
@@ -28,7 +34,7 @@ class Context {
    *
    */
-  repo (object) {
+  repo<T> (object: T) {
```
object is optional here, it can be undefined
replace with `repo<T> (object?: T)`

ocombe commented on src/context.ts:
```diff
@@ -49,11 +55,11 @@ class Context {
    * @param {object} [object] - Params to be merged with the issue params.
    */
-  issue (object) {
+  issue<T> (object: T) {
```
same here, object is optional

ocombe commented on src/context.ts:
```diff
@@ -9,8 +10,13 @@ const yaml = require('js-yaml')
+  id?: string
+  github: GitHubApi
+  log: LoggerWithTarget
+  payload!: WebhookPayloadWithRepository
```
I think that it should probably be something like `WebhookPayloadWithRepository<T>` so that we can pass the type of Github payload that it contains

philschatz replied:
:+1: for adding generics to the payload. I am not sure how that would work with `robot.on(eventName, handler)` but I think you can specify something like:

```typescript
type PushPayload = { /* ... */ }
type IssuesPayload = { /* ... */ }

interface IRobot {
  on(eventName: 'push',   callback: ({payload: PushPayload})   => void)
  on(eventName: 'issues', callback: ({payload: IssuesPayload}) => void)
}
```

ocombe commented on src/index.ts:
```diff
@@ -1,11 +1,14 @@
+import {logger} from './logger'
+import {createWebhookProxy} from './webhook-proxy'
+import * as express from 'express'
+import * as Logger from 'bunyan'
```
remove this import, you should use this here

tcbyrd replied:
@ocombe I ended up keeping this and just doing `export { Logger }` to get the compiler to stop warning me that I wasn't naming the export:

https://github.com/probot/probot/blob/920d5b67f64d54d471302836f3d59a40814f604c/src/index.ts#L116

Does that make sense here? I couldn't get any of the other workaround we discussed to work.

ocombe commented on src/index.ts:
```diff
@@ -77,7 +80,7 @@ module.exports = (options = {}) => {
     webhook,
     receive,
-    logger,
+    logger: Logger,
```
change this back to `logger,`, you want to export the probot logger, not bunyan itself

tcbyrd commented:
Adding `sudo: required` (in https://github.com/probot/probot/pull/372/commits/881675f86891eff53e8655e6bb8b6b1b67c3e3bb) allowed the build to finish and all tests pass. Need to work on fixing the new merge conflicts and updating for 6.0. Should be ready for integration tests after that.

tcbyrd commented:
Merge conflicts have been addressed and I published this to npm @ [probot-ts v 6.0.0-typescript](https://www.npmjs.com/package/probot-ts)

There's one tiny typescript compiler error that's causing the build to fail, related to type definitions that are being generated from Octokit: https://github.com/octokit/rest.js/pull/732#issuecomment-371619270, but `tsc` still emits the output even with that error.

If anyone reading this is interested in helping, I'd ❤️ love ❤️  if you can do a quick integration test by updating your probot dependency to [`probot-ts`](https://www.npmjs.com/package/probot-ts). Since `probot-run` looks at the compiled output in `lib/`, things _should_ work exactly the same.

bkeepers commented on package.json:
```diff
@@ -1,30 +1,53 @@
     "semantic-release": "semantic-release pre && npm publish && semantic-release post",
-    "postsemantic-release": "script/publish-docs"
+    "postsemantic-release": "script/publish-docs",
+    "prepublish": "npm run build"
```
When I run this I get:

```
npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.
```

philschatz replied:
What if `npm run build` was executed during `prepack` and `pretest` rather than `prepublish`? 

That way the TypeScript files would be compiled right before publishing to npm and right before tests ran.

https://docs.npmjs.com/misc/scripts has more on the various scripts options.

szeck87 replied:
I would just change it to `prepare`, I did that on atom-beautify.  They're the same thing as noted here:  https://docs.npmjs.com/misc/scripts#deprecation-note

bkeepers commented on src/context.ts:
```diff
@@ -9,8 +10,13 @@ const yaml = require('js-yaml')
-class Context {
-  constructor (event, github, log) {
+export class Context {
+  id?: string
```
Should ID ever be absent in real-world usage?

tcbyrd replied:
Was this removed? I get this error if I remove the`?` it because it's not defined anywhere else:

`Property 'id' has no initializer and is not definitely assigned in the constructor.`

I noticed it's not present in `context.js` https://github.com/probot/probot/blob/a0f84205f87cfebd15aa863b8f6cd84bda661f86/lib/context.js#L12-L16

tcbyrd replied:
I would think not. Can we then define it in the constructor instead? TS complains when it's not optional: `Property 'id' has no initializer and is not definitely assigned in the constructor.`

So we could do:
```js
constructor (event:any, github:GitHubApi, log:LoggerWithTarget) {
    Object.assign(this, event)
    this.github = github
    this.log = log
    this.id = event.id
  }
```

shiftkey commented on appveyor.yml:
```diff
@@ -6,6 +6,7 @@ environment:
 install:
   # Get the latest stable version of Node.js or io.js
   - ps: Install-Product node $env:nodejs_version
+  - npm install -g typescript
```
:thumbsup: just embed the version of `typescript` in your `package.json` and that'll get used in your build scripts

tcbyrd commented on package.json:
```diff
@@ -1,30 +1,50 @@
   "dependencies": {
     "@octokit/rest": "15.1.7",
-    "bottleneck": "^2.0.0",
+    "bottleneck": "^1.16.0",
```
Yep. Rolled back because 2.0 is not supporting ES5 syntax, which breaks compatibility with current FaaS providers. We can move it back once that's not an issue.

shiftkey commented on src/github.ts:
```diff
@@ -50,4 +52,35 @@ function EnhancedGitHubClient (options) {
 }
 
-module.exports = EnhancedGitHubClient
+export interface Options extends Octokit.Options {
```
Not sure how important this is, but using `extends` here means that you need to specify these members _and everything that `Octokit.Options` expects_, which is why the build is currently failing.

I think `baseUrl` maps to `host` in Octokit, and `agent` is the other required member for `Octokit.Options` which comes from  [`http.Agent`](https://nodejs.org/dist/latest-v8.x/docs/api/http.html#http_class_http_agent).

philschatz replied:
The `agent` was incorrectly marked as required but was fixed in https://github.com/octokit/rest.js/pull/803 (in `v15.2.3`).

shiftkey commented on src/index.ts:
```diff
@@ -28,29 +31,29 @@ module.exports = (options = {}) => {
     cert: options.cert
   })
-  const server = createServer({webhook, logger})
+  const server: express.Application = createServer({webhook, logger})
```
`: express.Application` should be inferred here, which means you can drop the `import * as express from 'express'` above

tcbyrd replied:
Yeh this one had me scratching my head. When i take it away, I get this:

`Exported variable 'createProbot' has or is using name 'e.Application' from external module "/Users/tcbyrd/ts-practice/probot/node_modules/@types/express/index" but cannot be named.`

So {createServer} is importing `express.Application`, but it gets aliased. I'm obviously missing something about how these types are being imported and exported.

shiftkey replied:
@tcbyrd yeah, let me wrap my head around what The Right Thing™ here is to do

codebytere commented:
how much work is left to do before this can potentially be merged? Anything I can do to help it along?

tcbyrd commented on package.json:
```diff
@@ -1,30 +1,50 @@
+    "build": "rimraf lib && tsc -p tsconfig.json",
     "start": "node ./bin/probot run",
-    "test": "jest --coverage && standard && npm run doc-lint",
+    "test": "jest --coverage --forceExit && standard && npm run doc-lint",
```
This can be cleaned up. It was part of me troubleshooting Travis errors.

tcbyrd commented:
@codebytere It's almost there. We had plans to release a beta tonight, but need to make some updates to bring it in line with the latest Octokit being used in `master`. If you know TypeScript, feel free to look over the open questions in the above reviews. https://github.com/probot/probot/pull/372#pullrequestreview-104407292 https://github.com/probot/probot/pull/372#discussion_r174964926

Additionally, if you have any Probot Apps and want to do some integration tests, you can update your `probot` dependency to [`probot-ts`](https://www.npmjs.com/package/probot-ts) and run your tests to see if anything is broken. I've tested against a few of the examples and the template, and a few others are using that to run their apps in serverless environments, but it would be 💖 for others to run some tests as well.

Once we get these last few things ironed out, we'll make an announcement.

codebytere commented:
@tcbyrd rad, we'll try and test it out in https://github.com/codebytere/trop!

bkeepers commented on src/github.ts:
```diff
@@ -50,4 +52,35 @@ function EnhancedGitHubClient (options) {
+  status: string
+}
+
+export interface OctokitWithPagination extends Octokit {
```
Could this add `paginate` to the existing `Octokit` interface instead of declaring a new interface?

bkeepers commented on src/github.ts:
```diff
@@ -50,4 +52,35 @@ function EnhancedGitHubClient (options) {
+  method: string
+  url: string
+  headers: any
+}
```
Can this and the next two interfaces be declared in the upstream octokit/rest.js repo?

bkeepers commented on src/plugins/stats.ts:
```diff
@@ -55,3 +52,17 @@ module.exports = async robot => {
+interface Repository {
+  private: boolean,
+  stargazers_count: number
+}
```
Is there a TypeScript project that declares types for GitHub API responses?

JasonEtco replied:
Wouldn't that be part of `@octokit/rest`? Since the return values from methods are pre-defined. @gr2m I can try to help with those, since it'd be quite the task to get em all (although maybe [this might be a great start](https://transform.now.sh/json-to-ts-interface/).

gr2m replied:
@philschatz started working it at https://github.com/octokit/rest.js/pull/732. I’m reluctant of having big JSON files describing all the REST API responses without having it some automation in place that makes sure these stay up-to-date

My current idea is to use https://github.com/gr2m/octokit-rest-routes to feed automated fixtures generation for all endpoints using https://github.com/octokit/fixtures and then use the recorded, normalised fixtures to extract types.

The challenge is that response change based on requested media type.

And at least in on case the response can differ based on the request path: https://developer.github.com/v3/git/refs/#get-a-reference. I try to work around this one on Octokit level though

philschatz replied:
The response types in https://github.com/octokit/rest.js/pull/732 were created by using auto-generated tests and fixtures for `octokat.js` but if you want to help getting them implemented in `gr2m/octokit-rest-routes` there is some discussion started in https://github.com/gr2m/octokit-rest-routes/issues/20

bkeepers commented on src/robot.ts:
```diff
@@ -160,7 +169,26 @@ class Robot {
+
+export interface EventWithEventField {
+  event: string
+}
```
Even though we don't use them, should this include all the properties from a webhook?  https://github.com/rvagg/github-webhook-handler/blob/master/github-webhook-handler.js#L92-L99

bkeepers commented on src/robot.ts:
```diff
@@ -160,7 +169,26 @@ class Robot {
-module.exports = (...args) => new Robot(...args)
+export const createRobot = (options: RobotOptions) => new Robot(options)
+
+export interface EventWithEventField {
```
Maybe `WebhookEvent` would be a better name here?

bkeepers commented on src/robot.ts:
```diff
@@ -160,7 +169,26 @@ class Robot {
+  event: string
+}
+
+// The TypeScript definition for cache-manager does not export the Cache interface so we recreate it here
```
It looks like there is an interface for this: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/cache-manager/index.d.ts

tcbyrd replied:
There is, but it's not exported. We could get out of sync if they change their function signature, but also that should just create a compiler error. Thoughts on leaving it for now?

bkeepers commented on src/serializers.ts:
```diff
@@ -46,3 +49,13 @@ module.exports = {
+  }
+}
+
+interface ExpressResponseWithDuration extends express.Response {
```
Can this be added directly to the Express.Response interface? Similar to this? https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express-session/index.d.ts#L17-L20

tcbyrd replied:
It could, but it seems simpler this way to me since we aren't exporting it. You're more familiar with types though. What benefits do you think it would add?

bkeepers commented on src/server.ts:
```diff
@@ -0,0 +1,22 @@
+  return app
+}
+
+export interface ServerArgs {webhook: express.Application, logger: Logger}
```
TypeScript best-practice question…

When do you export an interface for function parameters vs just declaring them inline? Could the `createServer` function above be declared like this?

```ts
export const createServer = function ({webhook: express.Application, logger: Logger}) {
```

tcbyrd replied:
In this instance, if you try and define it inline, everything in between `{}` gets treated as key/value pairs, but `express.Application` is a type, not a value. One of the things I've struggled with is that things between the brackets are sometimes objects and sometimes interfaces. This may be a good reason for defining your types in separate files, honestly. It may help differentiate between the two.

bkeepers commented on src/webhook-proxy.ts:
```diff
@@ -0,0 +1,23 @@
+      target: `http://localhost:${opts.port}${opts.path}`,
+      logger: opts.logger
+    })
+    return smee.start()
```
I'm not seeing type declarations for this. Where do they come from?

philschatz replied:
I believe `tsc` did not complain about missing definitions probably because the return value of `createWebhookProxy` is never used (see https://github.com/probot/probot/blob/typescript-init/src/index.ts#L89-L96 ).

urish commented:
Continuing the discussion from urish/typewiz#35. You mentioned you are almost there - can you give a one liner on what is missing?

tcbyrd commented:
Hey @urish! Thanks for the offer to help. We're mainly missing a few things from the GitHub API/Webhooks. See discussion [here](https://github.com/probot/probot/pull/372#discussion_r175258544) and [here](https://github.com/probot/probot/pull/372#discussion_r175258374). I think the dream would we something that ties into @gr2m's work in https://github.com/octokit/routes to generate types in sync with when GitHub adds routes to its documentation.

urish commented:
@tcbyrd so is your own code fully typed now, and you are just missing the typings for the GH APIs?

szeck87 commented:
Some suggestions:

* Consider using TSLint for linting. There are a couple of rule directories you can add such as `tslint-microsoft-contrib`
* Add `types` or `typings` to package.json (https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package)
* You should be able to just use `"ts-jest"` instead of the entire path in your jest config

ocombe commented on src/github/index.ts:
```diff
@@ -0,0 +1,67 @@
+ * @typedef github
+ * @see {@link https://github.com/octokit/rest.js}
+ */
+export const EnhancedGitHubClient = function (options) {
```
`function (options)` --> `function (options: Options)`

ocombe commented on src/robot.ts:
```diff
@@ -1,25 +1,34 @@
-class Robot {
-  constructor ({app, cache, router, catchErrors} = {}) {
+export class Robot {
+  public events: any
```
`public events: EventEmitter`

ocombe commented on src/index.ts:
```diff
@@ -19,7 +23,7 @@ const defaultApps = [
 ]
 
-module.exports = (options = {}) => {
+export const createProbot = (options: Options) => {
```
it would be very nice to have an interface as the returned type, so that we can get autocompletion for the different functions
I suggest:
```ts
export interface Probot {
  server: Server;
  webhook: Webhooks;
  receive: (event: WebhookEvent) => Promise<any>[];
  logger: Logger;
  load: (plugin: string | Plugin) => Robot;
  setup: (apps: Array<string | Plugin>) => void;
  start: () => void;
}

ocombe commented on src/server.ts:
```diff
@@ -0,0 +1,26 @@
+
+import {logRequest} from './middleware/logging'
+
+export const createServer = function (args: ServerArgs) {
```
return type: `express.Application`

ocombe commented on src/github/index.ts:
```diff
@@ -0,0 +1,67 @@
+}
+
+export interface OctokitWithPagination extends Octokit {
+  paginate: (res: Promise<Octokit.AnyResponse>, callback: (results: Octokit.AnyResponse) => void) => void
```
returns `Promise<any[]>`, not `void`

ocombe replied:
in fact it should probably be a generic type to avoid `any`:
```ts
export interface OctokitWithPagination<T> extends Octokit {
  paginate: (res: Promise<Octokit.AnyResponse>, callback: (results: Octokit.AnyResponse) => void) => Promise<T[]>
```

ocombe commented on src/plugins/stats.ts:
```diff
@@ -1,12 +1,9 @@
-module.exports = async robot => {
-  if (process.env.DISABLE_STATS) {
-    return
-  }
```
this should not have been removed, you probably did a wrong merge here

ocombe commented on src/github/graphql.ts:
```diff
@@ -1,18 +1,19 @@
 }
 
-async function query (octokit, query, variables = undefined, headers = {}) {
+async function query (octokit: OctokitWithPagination, query: string, variables: Variables, headers = {}) {
```
this function should allow generic types:
```ts

szeck87 commented on package.json:
```diff
@@ -1,30 +1,54 @@
+    "postsemantic-release": "script/publish-docs",
+    "prepublish": "npm run build"
   },
   "jest": {
```
I would consider moving this to a `jest.config.js` file in the root of the project.  package.json is already quite long.

tcbyrd commented:
@bkeepers I figured out why it wasn't running all the tests properly. The pattern for the test matcher wasn't looking at sub-directories (see 5b1dd99). This exposed a few minor things that needed to be updated. I then pulled the latest tests from `master` and updated with the latest PRs.

I feel pretty good about releasing this. There are some more improvements we can make, but now that we've got the existing tests passing, after the release I think we can work that out by gradually rewriting the tests in TypeScript.

LMK if you want to sync up to cut a release! 🎉

bkeepers commented:
TypeScript support is now released with the `next` tag:

```
$ npm install probot@next
+ probot@7.0.0-typescript.1
```

Try it out and let us know if you have any feedback or issues.

szeck87 commented:
Still need `typings` in package.json.

bkeepers commented:
@szeck87 thanks, should be included now in `7.0.0-typescript.2`

ocombe commented on src/wrap-logger.ts:
```diff
@@ -0,0 +1,48 @@
+  return fn
+}
+
+export interface LoggerWithTarget extends Logger {
```
You also need to add a callable signature so that people can do `this.robot.log('...')`.
This callable signature would work: `(str: string): void;`

ocombe replied:
The complete interface should be:

ocombe commented on src/github/graphql.ts:
```diff
@@ -1,18 +1,19 @@
 
-function addGraphQL (octokit) {
+export const addGraphQL = function (octokit) {
   octokit.query = query.bind(null, octokit)
```
because you're binding this on octokit, whenever I try to use `context.github.query`, typescript complains because `query` doesn't exist on the type `GithubAPI`. You should augment the typescript definition of the GithubApi to add this, or `context.github` should maybe be of type `GithubApiWithGraphQL` or something.

szeck87 replied:
Not sure why this is showed outdated here in the PR.  Definitely still an issue.

ocombe commented on src/context.ts:
```diff
@@ -9,9 +11,15 @@ const yaml = require('js-yaml')
+  public log: LoggerWithTarget
+  public payload!: WebhookPayloadWithRepository
+
+  constructor (event:any, github:GitHubApi, log:LoggerWithTarget) {
```
why do we need `log:LoggerWithTarget` here? it's never used

ocombe replied:
Also the parameter "github" is of type `OctokitWithPagination`, not `GithubApi`

ocombe commented on src/context.ts:
```diff
@@ -97,7 +105,7 @@ class Context {
    * @return {Promise<Object>} - Configuration object read from the file
    */
-  async config (fileName, defaultConfig) {
+  public async config<T> (fileName: string, defaultConfig: T) {
```
should defaultConfig be optional?

bkeepers replied:
Yep, want to open a PR for it? :grin:

ocombe commented on src/context.ts:
```diff
@@ -117,4 +125,29 @@ class Context {
 }
 
-module.exports = Context
+export interface PayloadRepository {
```
this interface is missing a lot of attributes that exist on some payloads, which makes typescript complains about non existent properties :(
eg: `id` is almost always available, so is `html_url`, ...
this makes this type unusable in most of the cases, and I have to override it which is super annoying
all the properties should be available here, and those that are not always present should be conditional (with a `?`).

ocombe replied:
Another solution is to add `[key: string]: any;` to each property so that it can include any other property and we don't have to define all of them, but only the ones defined here will be available for auto complete.
Or maybe use a generic type `<T>`?

bkeepers replied:
Yeah, I think we're going to need types for GitHub webhook payloads.

tcbyrd replied:
@ocombe I was going through some possibilities here and wanted to get your thoughts. The problem I see with trying to map types for the webhook payloads is that the schema is not consistent. Even for a given webhook `event`, the `action` might change the schema.

Example:

### IssuesEvent
Triggered when an issue is assigned, unassigned, labeled, unlabeled, opened, edited, milestoned, demilestoned, closed, or reopened.

key | type | description
-- | -- | --
`changes` | `object` | The changes to the issue **if the action was "edited"**.

So if you have:

`robot.on('issues.edited', context => {..}`

`context.payload` will have the `changes` object, but otherwise it will not.

Would it be acceptable to just stop at defining `context.payload` as any Object (`any{}`)? I don't know that even with the new [Conditional Types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html), we can define all the potential payload types without the type definition being overly complex.

I'm definitely not against trying to do that, I just don't know if the effort is necessary. Thoughts?

tcbyrd replied:
Ah, I like the first one! I'll go through the webhook event types to see how to define the  "generic" form, but then I think it would make the most sense to have a type for each Event, with optional types inside it:

`robot.on('issues.created', (context: IssuesContext) => {...}`
`robot.on('issue_comment.created', (context: IssueCommentContext) => {...}`

Then each Event Type Context would have a `changes?` key.

szeck87 commented:
Something happened with release 3 or 4, when I compile with `tsc`, it tries to include definition files under probot, even with node_modules excluded.  I have to include "skipLibCheck" in my tsconfig to get around it, but shouldn't be happening.

```
node_modules/probot/lib/github/index.d.ts(1,23): error TS2688: Cannot find type definition file for 'bunyan'.
node_modules/probot/lib/github/index.d.ts(41,15): error TS7006: Parameter 'OctokitRequestOptions' implicitly has an 'any' type.
node_modules/probot/lib/github/logging.d.ts(1,23): error TS2688: Cannot find type definition file for 'bunyan'.
node_modules/probot/lib/github/logging.d.ts(2,25): error TS7016: Could not find a declaration file for module 'bunyan'. '/Users/stevenzeck/Desktop/Development/issue-complete/node_modules/bunyan/lib/bunyan.js' implicitly has an 'any' type.
  Try `npm install @types/bunyan` if it exists or add a new declaration (.d.ts) file containing `declare module 'bunyan';`
node_modules/probot/lib/index.d.ts(1,23): error TS2688: Cannot find type definition file for 'express'.
node_modules/probot/lib/index.d.ts(2,23): error TS2688: Cannot find type definition file for 'bunyan'.
node_modules/probot/lib/index.d.ts(3,25): error TS7016: Could not find a declaration file for module 'bunyan'. '/Users/stevenzeck/Desktop/Development/issue-complete/node_modules/bunyan/lib/bunyan.js' implicitly has an 'any' type.
  Try `npm install @types/bunyan` if it exists or add a new declaration (.d.ts) file containing `declare module 'bunyan';`
node_modules/probot/lib/index.d.ts(4,26): error TS7016: Could not find a declaration file for module 'express'. '/Users/stevenzeck/Desktop/Development/issue-complete/node_modules/express/index.js' implicitly has an 'any' type.
  Try `npm install @types/express` if it exists or add a new declaration (.d.ts) file containing `declare module 'express';`
node_modules/probot/lib/robot.d.ts(1,23): error TS2688: Cannot find type definition file for 'express'.
node_modules/probot/lib/robot.d.ts(2,26): error TS7016: Could not find a declaration file for module 'express'. '/Users/stevenzeck/Desktop/Development/issue-complete/node_modules/express/index.js' implicitly has an 'any' type.
  Try `npm install @types/express` if it exists or add a new declaration (.d.ts) file containing `declare module 'express';`
node_modules/probot/lib/robot.d.ts(5,30): error TS7016: Could not find a declaration file for module 'promise-events'. '/Users/stevenzeck/Desktop/Development/issue-complete/node_modules/promise-events/emitter.js' implicitly has an 'any' type.
  Try `npm install @types/promise-events` if it exists or add a new declaration (.d.ts) file containing `declare module 'promise-events';`
node_modules/probot/lib/wrap-logger.d.ts(1,23): error TS2688: Cannot find type definition file for 'bunyan'.
```

szeck87 commented on src/robot.ts:
```diff
@@ -1,11 +1,13 @@
+import {logger} from './logger'
+import {LoggerWithTarget, wrapLogger} from './wrap-logger'
+
+  import {EventEmitter} from 'promise-events'
```
I would write a declaration file and declare the `promise-events` module.

szeck87 commented on src/robot.ts:
```diff
@@ -149,19 +160,19 @@ class Robot {
-  async auth (id, log = this.log) {
-    const github = new GitHubApi({
-      debug: process.env.LOG_LEVEL === 'trace',
+  public async auth (id?: number, log = this.log) {
```
Can you add the return type here as well?  Looks like `Promise<OctokitWithPagination>`.

szeck87 commented on src/robot.ts:
```diff
@@ -89,38 +99,39 @@ class Robot {
-      event.forEach(e => this.on(e, callback))
-      return
-    }
+  public on (eventName: string | string[], callback: (context: Context) => void) {
```
Should this always return a value, even in the else?

bkeepers commented:
I've tried this branch out on a couple projects written in pure JS and it's working great. I know there's a lot of improvements to make it a good experience to write apps in TypeScript, but I feel like there's going to be a lot of iteration to get to that point.

I'm going to merge this into master and release 7.0, and then other improvements can be followup PRs.
